ui/src/hooks/useHidRpc.ts

```diff
@@ -0,0 +1,251 @@
+    (n >> 16) & 0xFF,
+    (n >> 8) & 0xFF,
+    (n >> 0) & 0xFF,
+  ]);
+};
+
+const fromInt8ToUint8 = (n: number) => {
+  if (n < -128 || n > 127) {
+    throw new Error(`Number ${n} is not within the int8 range`);
+  }
+
+  return (n >> 0) & 0xFF;
+};
+
+const keyboardLedStateMasks = {
+  num_lock: 1 << 0,
+  caps_lock: 1 << 1,
+  scroll_lock: 1 << 2,
+  compose: 1 << 3,
+  kana: 1 << 4,
+  shift: 1 << 6,
+}
+
+export const toKeyboardLedState = (s: number): KeyboardLedState => {
+  if (!withinUint8Range(s)) {
+    throw new Error(`State ${s} is not within the uint8 range`);
+  }
+
+  return {
+    num_lock: (s & keyboardLedStateMasks.num_lock) !== 0,
+    caps_lock: (s & keyboardLedStateMasks.caps_lock) !== 0,
+    scroll_lock: (s & keyboardLedStateMasks.scroll_lock) !== 0,
+    compose: (s & keyboardLedStateMasks.compose) !== 0, // TODO: check if this is correct
+    kana: (s & keyboardLedStateMasks.kana) !== 0,
+    shift: (s & keyboardLedStateMasks.shift) !== 0,
+  } as KeyboardLedState;
+};
+
+const toPointerReportEvent = (x: number, y: number, buttons: number) => {
+  if (!withinUint8Range(buttons)) {
+    throw new Error(`Buttons ${buttons} is not within the uint8 range`);
+  }
+
+  return new Uint8Array([
+    HID_RPC_MESSAGE_TYPES.PointerReport,
+    ...fromInt32toUint8(x),
+    ...fromInt32toUint8(y),
+    buttons,
+  ]);
+};
+
+const toMouseReportEvent = (dx: number, dy: number, buttons: number) => {
+  if (!withinUint8Range(buttons)) {
+    throw new Error(`Buttons ${buttons} is not within the uint8 range`);
+  }
+  return new Uint8Array([
+    HID_RPC_MESSAGE_TYPES.MouseReport,
+    fromInt8ToUint8(dx),
+    fromInt8ToUint8(dy),
+    buttons,
+  ]);
+};
+
+const toKeyboardReportEvent = (keys: number[], modifier: number) => {
+  if (!withinUint8Range(modifier)) {
+    throw new Error(`Modifier ${modifier} is not within the uint8 range`);
+  }
+
+  keys.forEach((k) => {
+    if (!withinUint8Range(k)) {
+      throw new Error(`Key ${k} is not within the uint8 range`);
+    }
+  });
+
+  return new Uint8Array([
+    HID_RPC_MESSAGE_TYPES.KeyboardReport,
+    modifier,
+    ...keys,
+  ]);
+};
+
+const toKeypressReportEvent = (key: number, press: boolean) => {
+  if (!withinUint8Range(key)) {
+    throw new Error(`Key ${key} is not within the uint8 range`);
+  }
+
+  return new Uint8Array([
+    HID_RPC_MESSAGE_TYPES.KeypressReport,
+    key,
+    press ? 1 : 0,
+  ]);
+};
+
+const toHandshakeMessage = () => {
+  return new Uint8Array([HID_RPC_MESSAGE_TYPES.Handshake]);
+};
+
+export interface HidRpcMessage {
+  type: HidRpcMessageType;
+  keysDownState?: KeysDownState;
+}
+
+const unmarshalHidRpcMessage = (data: Uint8Array): HidRpcMessage | undefined => {
+  if (data.length < 1) {
+    throw new Error(`Invalid HID RPC message length: ${data.length}`);
+  }
+
+  const payload = data.slice(1);
+
+  switch (data[0]) {
+    case HID_RPC_MESSAGE_TYPES.Handshake:
+      return {
+        type: HID_RPC_MESSAGE_TYPES.Handshake,
+      };
```

Comment:
Would we be better served to actually return a protocol version in the handshake for future use?